### ISSUE TITLE
Output file atomic update

### ISSUE THREAD

**Issue by nh2** [rui314/mold#188]
Dear maintainer,

I found that in [this Phoronix forum entry](https://www.phoronix.com/forums/forum/phoronix/latest-phoronix-articles/1297685-mold-1-0-released-as-a-modern-high-speed-linker-alternative-to-gnu-gold-llvm-lld/page3#post1297869) it was discussed that mold does not do atomic writes, that is, that it overwrites binaries in place (instead of create new + rename).

Is that correct?

If yes, could atomic file writing support be added?

It is easy to implement and has various benefits; beyond solving the `text file busy` problem, it also makes it impossible that when the disk is full or the computer powers off unexpectedly, a build system writes a half-written binary (which in case of mtime-based build systems like `make` would result in the build system not detecting that a rebuild is necessary).

Thanks!

---

Side note:

I've had a short look at https://github.com/rui314/mold/blob/733bb6354f9913a9fba599f1779620286f1e2e4c/elf/output-file.cc which suggests that some atomic writes (`rename()`) are in place. What I don't know yet:

* Are there other non-atomic file writes in the code that are relevant?
* What is the purpose of the `rename(path.c_str(), output_tmpfile)` that happens at the beginning of `MemoryMappedOutputFile`, moving the existing file to a temporary location? As per [`design.md`](https://github.com/rui314/mold/blob/56f63dd295204b4845c898ed25eb212551a8063b/docs/design.md), incremental linking is not used, so I'm wondering what is done with the copied existing file.

**Comment by rui314** [comment]
Here is what mold is doing:

1. Rename an existing output file a temporary file
2. Overwrite to the temporary file
3. Rename the temporary file the output file name

The purpose of doing this is, if something goes wrong in step 2, only a temporary file is left. No corrupted file is left as an output file. So, if mold fails to write to a file for some reason, there could be two outcomes:

1. Nothing happens, or
2. Existing output file is gone (renamed a temporary file)

The above outcome is better than leaving a corrupted file as an output file.

**Comment by nh2** [comment]
@rui314 Thanks for the quick answer! That means that mold already does atomic writes for all files it writes, right?

I guess what still confuses me is why (1) is necessary. Since you never use the existing output file for anything, it seems superfluous.

Could you not simply:

1. [removed]
2. Write output to a new temporary file
3. Rename the temporary file the output file name

?

**Comment by rui314** [comment]
We use the existing file. In the first step, we rename an existing file, and in the second step, we overwrite to *the renamed* file.

**Comment by nh2** [comment]
@rui314 I mean, are you using the existing file's **content** for anything? It doesn't look like it, so my question is what the benefit of rename+open+ETXTBSY-handling+rename is vs just a simpler open+rename.

**Comment by rui314** [comment]
No, we are not using any _content_ of an existing file for anything. And if an existing file is in use and mold gets ETXTBSY, there's indeed no benefit of attempting to use the existing file.

But if the existing file is not in use, writing to it is noticeably faster than creating a new file and writing to it.

**Comment by nh2** [comment]
> But if the existing file is not in use, writing to it is noticeably faster than creating a new file and writing to it.

Ah, interesting. I wasn't aware of the fact that writing to existing files can be faster than writing to new ones.

Thanks for answering my questions!

**Comment by nh2** [comment]
@rui314 Actually, one more question from https://www.phoronix.com/forums/forum/software/programming-compilers/1297685-mold-1-0-released-as-a-modern-high-speed-linker-alternative-to-gnu-gold-llvm-lld/page3#post1297869

It claims:

> open fails only for running executables, but not for their libraries, overwriting used library will certainly crash running executables using it

I understand this as the user saying: The `ETXTBSY` safety mechanism only works for executables, but if `mold` writes to an `.so` files in-place, it can crash existing running programs that use that `.so` file.

Do you know if that is true?

**Comment by rui314** [comment]
@nh2 I think what that user said is correct. I don't know how the runtime behaves if a .so in use is overwritten, but if it could results in a bad outcome (such as a process crash), mold should open a new file when creating a .so file to be on the safe side. Let me take a look.

**Comment by llunak** [comment]
To repeat the info from #517, on Linux, open() fails with ETXTBUSY only for executables, but not for shared libraries (https://lwn.net/Articles/866493/). Re-linking a shared library used by a running process crashes.

I've tried with a testcase [moldlinktest.zip](https://github.com/rui314/mold/files/8697157/moldlinktest.zip) (adjust Makefile, a.h and 'make') and indeed overwriting is faster.

So maybe allow this only without -shared, and additionally also if some getenv() is set? And I have no idea what the ETXTBUSY situation is for other OSes (and quick search doesn't bring up anything).

**Comment by llunak** [comment]
Oh, and I have no idea if there's another cheap way to detect if a shared library is used by a running process. I doubt there is, since AFAICT ETXTBUSY is meant to be that. Given that mold is now getting big enough, I suppose you could try asking some kernel developers if they can either fix that or do something about the performance.

**Comment by rui314** [comment]
I think from the kernel's point of view, shared object files are just mmap'ed files. Maybe libc can in theory do something about that, but it's unlikely in practice.

**Comment by llunak** [comment]
The LWN article has details on that (including comments below saying that it doesn't seem to make sense to be different for executables and libraries, including suggestions about possible ways of how that could be enabled again).

**Comment by rui314** [comment]
Thanks. I read that article, and it feels like the article were written exactly for us. mold is a "crazy user-land program" that depends on ETXTBSY. It may be worth to bring this topic to the Linux kernel mailing list.

One thing to note is that we don't actually _need_ to write to an existing file. That's a hack to workaround a slowness of creating a new file. That's probably the issue that the kernel has to solve rather than supporting ETXTBSY for .so files.

**Comment by nh2** [comment]
> If creating a new file and writing to it is as fast as writing to an existing file

@rui314 I recently painfully learned this following quirks:

On ext4 specifically, overwriting an existing file is actually much slower than deleting the existing file and then writing it anew.

This is surprising at first, since one would expect that writing pre-allocated files would save the overhead of allocating the space.

But it is because ext4 inserts a hidden `fsync()` into the `close()` operation based on certain heuristics. This was the "solution" to the https://en.wikipedia.org/wiki/Ext4#Delayed_allocation_and_potential_data_loss issue a couple years ago.

From https://www.kernel.org/doc/Documentation/filesystems/ext4.txt (prettier render [here](https://man7.org/linux/man-pages/man5/ext4.5.html)):

> If `auto_da_alloc` is enabled, ext4 will detect the replace-via-rename and replace-via-truncate patterns and force that any delayed allocation blocks are allocated

and `auto_da_alloc` is the default.

I learned that when I tried a parallel `unzip` library written in Rust; unpacking the files was 10x faster in the first run than in the second run (which overrode the existing files with `O_TRUNC`).

Both `O_TRUNC` and atomic-`rename()` are detected by that heuristic.

The only way to not have an implicit fsync hidden in `close()` on modern ext4 is delete the existing files first, or `rename()` them out of the way.

Other file systems don't have that problem; they continue to assume that you fsync when you want fsync semantics.
